src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,447 @@
+                Debug.Assert((stat.Mode & (int)forbiddenPermissions) == 0, "Created PFX has too broad of permissions");
+#endif
+            }
+
+            // Null out _certificates so the next call to get_Certificates causes a re-scan.
+            _certificates = null;
+        }
+
+        public void Remove(ICertificatePal certPal)
+        {
+            OpenSslX509CertificateReader cert = (OpenSslX509CertificateReader)certPal;
+
+            using (X509Certificate2 copy = new X509Certificate2(cert.DuplicateHandles()))
+            {
+                bool hadCandidates;
+                string currentFilename = FindExistingFilename(copy, _storePath, out hadCandidates);
+
+                if (currentFilename != null)
+                {
+                    if (_readOnly)
+                    {
+                        // Windows compatibility, the readonly check isn't done until after a match is found.
+                        throw new CryptographicException(SR.Cryptography_X509_StoreReadOnly);
+                    }
+
+                    File.Delete(currentFilename);
+                }
+            }
+
+            // Null out _certificates so the next call to get_Certificates causes a re-scan.
+            _certificates = null;
+        }
+
+        private static string FindExistingFilename(X509Certificate2 cert, string storePath, out bool hadCandidates)
+        {
+            hadCandidates = false;
+
+            foreach (string maybeMatch in Directory.EnumerateFiles(storePath, cert.Thumbprint + PfxWildcard))
+            {
+                hadCandidates = true;
+
+                try
+                {
+                    using (X509Certificate2 candidate = new X509Certificate2(maybeMatch))
+                    {
+                        if (candidate.Equals(cert))
+                        {
+                            return maybeMatch;
+                        }
+                    }
+                }
+                catch (CryptographicException)
+                {
+                    // Contents weren't interpretable as a certificate, so it's not a match.
+                }
+            }
+
+            return null;
+        }
+
+        private string FindOpenSlot(string thumbprint)
+        {
+            // We already know that {thumbprint}.pfx is taken, so start with {thumbprint}.1.pfx
+
+            // We need space for {thumbprint} (thumbprint.Length)
+            // And ".0.pfx" (6)
+            // If MaxSaveAttempts is big enough to use more than one digit, we need that space, too (MaxSaveAttempts / 10)
+            StringBuilder pathBuilder = new StringBuilder(thumbprint.Length + PfxOrdinalWildcard.Length + (MaxSaveAttempts / 10));
+
+            pathBuilder.Append(thumbprint);
+            pathBuilder.Append('.');
+            int prefixLength = pathBuilder.Length;
+
+            for (int i = 1; i <= MaxSaveAttempts; i++)
+            {
+                pathBuilder.Length = prefixLength;
```

Comment:
Nit (not worth fixing unless you're going to make additional changes): this could move to the end of the loop, since the first time in it's already the right length.